### PR TITLE
[7.4 only] LPS-123370 Create new clay:navigation-card tag

### PR DESCRIPTION
**Pull request by markocikos**
See https://issues.liferay.com/browse/LPS-123370 for task description. 

In this PR we are:
1. Creating new `<clay:navigation-card>` tag based on [Lexicon guidelines](https://liferay.design/lexicon/core-components/cards/cards-navigation/) and [Clay CardWithNavigation](https://github.com/liferay/clay/blob/master/packages/clay-card/src/CardWithNavigation.tsx)
1. Adding few samples in `frontend-taglib-clay-sample-web`

Notes:
- The tag HTML markup is equal to Clay component markup. There is only server-side rendering, there is no need for hydration.
- Starting point for API was Clay component API, but I made few changes:
  - Instead of `horizontal` prop, tag has the `small` attribute. This is to better match Lexicon's guideline to have a 'Small' and a 'Big' card.
  - Instead of `horizontalSymbol` prop, tag has the `icon` attribute. This is to be consistent with many other tags.
- There is a small style difference between sample-web (see screenshot) and [storybook](https://storybook.clayui.com/?path=/story/components-claycard--cardwithnavigation). On the big card, aspect ratio is different. It is caused by [this line](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-theme/frontend-theme-classic/src/css/custom_properties/_custom_properties_set.scss#L4) in frontend-theme. I'm ok with this.
- I was in two minds about adding support for custom content in tag. In Clay, we have support for custom content, but Lexicon does nor imply it. Lexicon states that card should have 'just a minimalistic image or icon'. This could be interpreted that we want an opinionated tag with `imageSrc` and `imageAlt` attributes, and fixed markup. In the end, I settled on custom content for extra flexibility and Clay markup consistency.

Screenshot, without noise of other cards:
screen

### CODE REVIEW

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by carloslancha** [comment]
Hey @markocikos can we put this one on hold until we merge https://github.com/liferay-frontend/liferay-portal/pull/498 so we can update this one to use `BaseCardTag`?

**Comment by markocikos** [comment]
@julien @carloslancha force pushed, updated to use `BaseCardTag`. Open for review!

**Comment by markocikos** [comment]
ci:test:sf

**Review by julien — APPROVED** [review]

**Comment by markocikos** [comment]
ci:test:relevant

**Review by kresimir-coko — APPROVED** [review]
Not terrible ❤️

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:3345]: Based on Lexicon definition there will always be a title, don't know if we want to make it mandatory here

**Review by carloslancha — CHANGES_REQUESTED** [review]
As talked offline we have use cases where the navigation card click will be attached to some js stuff (like opening a modal), so we should hydrate the tag so we can accept a `propsTransformer` to do it.

That implies a problem with non small navigation cards, which I commented in #

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java:124]: > I was in two minds about adding support for custom content in tag. In Clay, we have support for custom content, but Lexicon does nor imply it. Lexicon states that card should have 'just a minimalistic image or icon'. This could be interpreted that we want an opinionated tag with imageSrc and imageAlt attributes, and fixed markup. In the end, I settled on custom content for extra flexibility and Clay markup consistency.

Well, with custom content being hard blocked, I guess I don't have to be in two minds anymore. I will need to add `imageSrc` and `imageAlt` to support images, per Lexicon guidelines.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/liferay-clay.tld:3345]: It does make sense to make it mandatory.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java:124]: Yup, makes sense 😄

**Comment by markocikos** [comment]
Self-review: I need to add `NavigationCard` support. It would be complex, as well as unwanted by product teams, to update model-based horizontal cards to attribute-based navigation cards..

**Comment by markocikos** [comment]
@carloslancha @julien Force pushed, ready for review! All changes are in `[WIP]` commit for hopefully easier review. If all is ok, I will merge to previous commits before forwarding.

Changes:
- Add model object, `NavigationCard`. Apply attributes hierarchy consistent to other cards. Add a sample.
- Add hydration, as a support for `propsTransformer`. Add a sample.
- Replace custom body support with fixed icon or image markup. This means adding `imageSrc` and `imageAlt` attributes.
- Move `href` setting to `doStartTag`, for consistency.
- No need to set 'div' as it is set by default.

Notes:
- I didn't set `title` to mandatory, as discussed in https://github.com/liferay-frontend/liferay-portal/pull/503#discussion_r525195785. This is because `title` can be set in the model object.
- I set default `isSmall` to `null` in the model object. This is because we had issues previously with opposite defaults in model, tag and Clay. Default is still set to `false` in the tag, to match default from Clay.
- The image in the last card (see screenshot) overflows container. It is, in fact, an elephant :slightly_smiling_face: Overflow may be intentional, left for devs to override if needed. If it is not, it should be fixed in Clay.
- There are quite a few trailing props from `BaseCardTag` that `NavigationCard` is not using and I had to filter out from props: `actions, disabled, inputName, inputValue, selectable, selected`. It's probably ok for now, but if we have another card that does not use those we should consider moving those non-common ones from base class.


screen

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
On hold, blocked by https://issues.liferay.com/browse/LPS-123844

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java:197]: What if both `icon` and `imageSrc` are defined? Will both be rendered?

**Review by carloslancha — COMMENTED** [review]
Just added one comment inline, the rest LGTM. Oh, ok, LNHTM.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java:197]: If the card is big and both are set, yes. If the card is small, only `icon` will be rendered. In this we are following the condition logic from [Clay component](https://github.com/liferay/clay/blob/master/packages/clay-card/src/CardWithNavigation.tsx), the only difference being instead of `{children}` we have the hardcoded ``.

I think this is fine, it's a minor edge case. Who knows, maybe there could even be an odd case where we would want icon superimposed on image, that does something different than navigation :man_shrugging:

**Comment by markocikos** [comment]
Self-review: there is no need for `_writeBody` as a separate method, that's a leftover from custom content support.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java
Comment (carloslancha) [reply modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java:197]: Taking a look to [Lexicon definition](https://liferay.design/lexicon/core-components/cards/cards-navigation/): 

>This card is simple containing just a minimalistic image or icon, title and possible short description.

It feels like only one of them should be rendered, it actually feels weird to me to render both. It's indeed an edge case (nobody should pass both attributes) but just in case I'd safe check it.

**Comment by markocikos** [comment]
@carloslancha  pushed, see `REVIEW-2` commit 

Changes:
- Image and icon can no longer show at the same time. If both are set in a big card, image has priority.
- Removed unnecessary `_writeBody` method
- In server side rendering, replaced direct vars with getters (for the case when both model and attributes are set)

It makes sense to me that image would have priority over icon, since that is the intended use. As this is a fringe case with obvious consequences and solution, I don't think we need to be defensive about it, like throwing error or console warning.

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/NavigationCardTag.java:197]: Sorting (imageSrc after icon)

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/NavigationCard.js
Comment (carloslancha) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/cards/NavigationCard.js:50]: Since this logic about rendering the children only when it's not horizontal is already defined in the component I guess we can simplify this to:

```
{imageSrc ? ( 
      <img alt={imageAlt} src={imageSrc} />
) : (
      symbol && <ClayIcon symbol={symbol} />
)}
```

**Review by carloslancha — APPROVED** [review]
Left just a copule of nits. LGTM.

**Comment by markocikos** [comment]
Added the last SF commit. Tomorrow, after NPE fix lands, I'll merge commits, rerun tests and forward.

**Comment by carloslancha** [comment]
@markocikos NPE fix it's on master already

**Comment by markocikos** [comment]
Haha must have landed few seconds after my comment

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
Failures seem unrelated. Rebased and pushed, no changes.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
Please don't forward this yet, even if it passes CI. It needs update, I'm working on it.

**Comment by markocikos** [comment]
@carloslancha @julien please review last two commits, [1](https://github.com/liferay-frontend/liferay-portal/pull/503/commits/f2bd96b4a6347abac197a5e1e41f4e1351117e04) and [2](https://github.com/liferay-frontend/liferay-portal/pull/503/commits/b241afe39a44ad812cd9c81c4471c1696ac69543).

File: modules/apps/frontend-taglib/frontend-taglib-clay-sample-web/package.json
Comment (julien) [inline modules/apps/frontend-taglib/frontend-taglib-clay-sample-web/package.json:7]: What's the reason for removing the `test` script? Is it useless?

**Review by carloslancha — APPROVED** [review]
LGTM!

File: modules/apps/frontend-taglib/frontend-taglib-clay-sample-web/package.json
Comment (markocikos) [reply modules/apps/frontend-taglib/frontend-taglib-clay-sample-web/package.json:7]: It was breaking CI. It tried to run these tests and there were none.

Comment (julien) [reply modules/apps/frontend-taglib/frontend-taglib-clay-sample-web/package.json:7]: Ok so the reason is because this has been copy pasted and we have no tests, so we don't need it.
This PR is not horrible :heart:

**Review by julien — APPROVED** [review]

**Comment by markocikos** [comment]
ci:forward

**Comment by carloslancha** [comment]
ci:forward

**Pull request closed without merging**
